Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp

```diff
@@ -0,0 +1,823 @@
+
+    bool isSupportedConditionOp(NodeType op);
+    bool isSupportedUpdateOp(NodeType op);
+
+    ComparisonFunction comparisonFunction(Node* condition);
+    UpdateFunction updateFunction(Node* update);
+
+    bool identifyInductionVariable(LoopData& data)
+    {
+        Node* condition = data.condition();
+        auto isConditionValid = [&]() ALWAYS_INLINE_LAMBDA {
+            if (!isSupportedConditionOp(condition->op()))
+                return false;
+
+            // Condition left
+            Node* update = condition->child1().node();
+            if (!isSupportedUpdateOp(update->op()))
+                return false;
+            // FIXME: Currently, we assume the left operand is the induction variable.
+            if (update->child1()->op() != GetLocal)
+                return false;
+            if (!update->child2()->isInt32Constant())
+                return false;
+
+            // Condition right
+            if (!condition->child2()->isInt32Constant())
+                return false;
+
+            data.operand = condition->child2()->asInt32();
+            data.update = condition->child1().node();
+            data.updateValue = update->child2()->asInt32();
+            data.inductionVariable = condition->child1()->child1().node();
+            return true;
+        };
+        if (!isConditionValid()) {
+            dataLogLnIf(Options::verboseLoopUnrolling(), "Skipping loop with header ", *data.header(), "since the invalid loop condition node D@", condition->index());
+            return false;
+        }
+
+        auto isInitialValueValid = [&]() ALWAYS_INLINE_LAMBDA {
+            Node* initialization = nullptr;
+            for (Node* n : *data.preHeader) {
+                if (n->op() != SetLocal || !data.isInductionVariable(n))
+                    continue;
+                initialization = n;
+            }
+            if (!initialization || !initialization->child1()->isInt32Constant())
+                return false;
+            data.initialValue = initialization->child1()->asInt32();
+            return true;
+        };
+        if (!isInitialValueValid()) {
+            dataLogLnIf(Options::verboseLoopUnrolling(), "Skipping loop with header ", *data.header(), " since the initial value is invalid");
+            return false;
+        }
+
+        auto isInductionVariableValid = [&]() ALWAYS_INLINE_LAMBDA {
+            uint32_t updateCount = 0;
+            for (uint32_t i = 0; i < data.loopSize(); ++i) {
+                BasicBlock* body = data.loopBody(i);
+                for (Node* node : *body) {
+                    if (node->op() != SetLocal || !data.isInductionVariable(node))
+                        continue;
+                    dataLogLnIf(Options::verboseLoopUnrolling(), "Induction variable ", data.inductionVariable->index(), " is updated at node ", node->index(), " at ", *body);
+                    ++updateCount;
+                    if (updateCount != 1)
+                        return false;
```

Comment:
Probably worth extending in the future.

Reply:
Yeah, that could be tricky for example 
```
for (let i = 0; i < 4: ) {
    ...
    if (...) {
         i += 1;
    } else {
         i += 2;
    }
}
```
this makes it difficult to extract a constant number of iterations of the loop. The reason why this simple loop unrolling with constant iterations works is that
1. We ensure only one update.
2. That update must be traceable/reachable from the terminal (branch) in the tail block.

And this one doesn't work since the terminal (`i < 4`) gets the induction variable from `Phi`.
```
for (let i = 0; i < 4: ) {
    ...
    if (...) {
         i += 1;
    }
}
```

Reply:
Both examples shown above make loop unrolling with constant iterations unpredictable.

Reply:
Added a FIXME there.